<!DOCTYPE html >
<html>
        <head>
          <title>clustering - Smile - Statistical Machine Intelligence and Learning Engine - smile.clustering</title>
          <meta name="description" content="clustering - Smile - Statistical Machine Intelligence and Learning Engine - smile.clustering" />
          <meta name="keywords" content="clustering Smile Statistical Machine Intelligence and Learning Engine smile.clustering" />
          <meta http-equiv="content-type" content="text/html; charset=UTF-8" />
          
      <link href="../../lib/template.css" media="screen" type="text/css" rel="stylesheet" />
      <link href="../../lib/diagrams.css" media="screen" type="text/css" rel="stylesheet" id="diagrams-css" />
      <script type="text/javascript" src="../../lib/jquery.js" id="jquery-js"></script>
      <script type="text/javascript" src="../../lib/jquery-ui.js"></script>
      <script type="text/javascript" src="../../lib/template.js"></script>
      <script type="text/javascript" src="../../lib/tools.tooltip.js"></script>
      
      <script type="text/javascript">
         if(top === self) {
            var url = '../../index.html';
            var hash = 'smile.clustering.package';
            var anchor = window.location.hash;
            var anchor_opt = '';
            if (anchor.length >= 1)
              anchor_opt = '@' + anchor.substring(1);
            window.location.href = url + '#' + hash + anchor_opt;
         }
   	  </script>
    
        </head>
        <body class="value">
      <div id="definition">
        <img alt="Package" src="../../lib/package_big.png" />
        <p id="owner"><a href="../package.html" class="extype" name="smile">smile</a></p>
        <h1>clustering</h1><span class="permalink">
      <a href="../../index.html#smile.clustering.package" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      </div>

      <h4 id="signature" class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <span class="name">clustering</span>
      </span>
      </h4>
      
          <div id="comment" class="fullcommenttop"><div class="comment cmt"><p>Clustering analysis. Clustering is the assignment of a set of observations
into subsets (called clusters) so that observations in the same cluster are
similar in some sense. Clustering is a method of unsupervised learning,
and a common technique for statistical data analysis used in many fields.</p><p>Hierarchical algorithms find successive clusters using previously
established clusters. These algorithms usually are either agglomerative
(&quot;bottom-up&quot;) or divisive (&quot;top-down&quot;). Agglomerative algorithms begin
with each element as a separate cluster and merge them into successively
larger clusters. Divisive algorithms begin with the whole set and proceed
to divide it into successively smaller clusters.</p><p>Partitional algorithms typically determine all clusters at once, but can
also be used as divisive algorithms in the hierarchical clustering.
Many partitional clustering algorithms require the specification of
the number of clusters to produce in the input data set, prior to
execution of the algorithm. Barring knowledge of the proper value
beforehand, the appropriate value must be determined, a problem on
its own for which a number of techniques have been developed.</p><p>Density-based clustering algorithms are devised to discover
arbitrary-shaped clusters. In this approach, a cluster is regarded as
a region in which the density of data objects exceeds a threshold.</p><p>Subspace clustering methods look for clusters that can only be seen in
a particular projection (subspace, manifold) of the data. These methods
thus can ignore irrelevant attributes. The general problem is also known
as Correlation clustering while the special case of axis-parallel subspaces
is also known as two-way clustering, co-clustering or biclustering in
bioinformatics: in these methods not only the objects are clustered but
also the features of the objects, i.e., if the data is represented in
a data matrix, the rows and columns are clustered simultaneously. They
usually do not however work with arbitrary feature combinations as in general
subspace methods.
</p></div><div class="toggleContainer block">
          <span class="toggle">Linear Supertypes</span>
          <div class="superTypes hiddenContent"><span class="extype" name="scala.AnyRef">AnyRef</span>, <span class="extype" name="scala.Any">Any</span></div>
        </div></div>
        

      <div id="mbrsel">
        <div id="textfilter"><span class="pre"></span><span class="input"><input id="mbrsel-input" type="text" accesskey="/" /></span><span class="post"></span></div>
        <div id="order">
              <span class="filtertype">Ordering</span>
              <ol>
                
                <li class="alpha in"><span>Alphabetic</span></li>
                <li class="inherit out"><span>By Inheritance</span></li>
              </ol>
            </div>
        <div id="ancestors">
                <span class="filtertype">Inherited<br />
                </span>
                <ol id="linearization">
                  <li class="in" name="smile.clustering"><span>clustering</span></li><li class="in" name="scala.AnyRef"><span>AnyRef</span></li><li class="in" name="scala.Any"><span>Any</span></li>
                </ol>
              </div><div id="ancestors">
            <span class="filtertype"></span>
            <ol>
              <li class="hideall out"><span>Hide All</span></li>
              <li class="showall in"><span>Show All</span></li>
            </ol>
          </div>
        <div id="visbl">
            <span class="filtertype">Visibility</span>
            <ol><li class="public in"><span>Public</span></li><li class="all out"><span>All</span></li></ol>
          </div>
      </div>

      <div id="template">
        <div id="allMembers">
        

        

        

        <div id="values" class="values members">
              <h3>Value Members</h3>
              <ol><li name="smile.clustering.$dummy" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="$dummy"></a>
      <a id="$dummy:$dummy"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">object</span>
      </span>
      <span class="symbol">
        <a href="package$$$dummy$.html"><span class="name">$dummy</span></a>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@$dummy" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">Hacking scaladoc <a href="https://github.com/scala/bug/issues/8124" target="_blank">issue-8124</a>.</p><div class="fullcomment"><div class="comment cmt"><p>Hacking scaladoc <a href="https://github.com/scala/bug/issues/8124" target="_blank">issue-8124</a>.
The user should ignore this object.</p></div></div>
    </li><li name="smile.clustering#clarans" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="clarans[T&lt;:AnyRef](data:Array[T],distance:smile.math.distance.Distance[T],k:Int,maxNeighbor:Int,numLocal:Int):smile.clustering.CLARANS[T]"></a>
      <a id="clarans[T&lt;:AnyRef](Array[T],Distance[T],Int,Int,Int):CLARANS[T]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">clarans</span><span class="tparams">[<span name="T">T &lt;: <span class="extype" name="scala.AnyRef">AnyRef</span></span>]</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="smile.clustering.clarans.T">T</span>]</span>, <span name="distance">distance: <span class="extype" name="smile.math.distance.Distance">Distance</span>[<span class="extype" name="smile.clustering.clarans.T">T</span>]</span>, <span name="k">k: <span class="extype" name="scala.Int">Int</span></span>, <span name="maxNeighbor">maxNeighbor: <span class="extype" name="scala.Int">Int</span></span>, <span name="numLocal">numLocal: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">16</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.CLARANS">CLARANS</span>[<span class="extype" name="smile.clustering.clarans.T">T</span>]</span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@clarans[T&lt;:AnyRef](data:Array[T],distance:smile.math.distance.Distance[T],k:Int,maxNeighbor:Int,numLocal:Int):smile.clustering.CLARANS[T]" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">Clustering Large Applications based upon RANdomized Search.</p><div class="fullcomment"><div class="comment cmt"><p>Clustering Large Applications based upon RANdomized Search. CLARANS is an
efficient medoid-based clustering algorithm. The k-medoids algorithm is an
adaptation of the k-means algorithm. Rather than calculate the mean of the
items in each cluster, a representative item, or medoid, is chosen for each
cluster at each iteration. In CLARANS, the process of finding k medoids from
n objects is viewed abstractly as searching through a certain graph. In the
graph, a node is represented by a set of k objects as selected medoids. Two
nodes are neighbors if their sets differ by only one object. In each iteration,
CLARANS considers a set of randomly chosen neighbor nodes as candidate
of new medoids. We will move to the neighbor node if the neighbor
is a better choice for medoids. Otherwise, a local optima is discovered. The
entire process is repeated multiple time to find better.</p><p>CLARANS has two parameters: the maximum number of neighbors examined
(maxNeighbor) and the number of local minima obtained (numLocal). The
higher the value of maxNeighbor, the closer is CLARANS to PAM, and the
longer is each search of a local minima. But the quality of such a local
minima is higher and fewer local minima needs to be obtained.</p><h6>References:</h6><ul><li>R. Ng and J. Han. CLARANS: A Method for Clustering Objects for Spatial Data Mining. IEEE TRANS. KNOWLEDGE AND DATA ENGINEERING, 2002.
</li></ul></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>the data set.</p></dd><dt class="param">distance</dt><dd class="cmt"><p>the distance/dissimilarity measure.</p></dd><dt class="param">k</dt><dd class="cmt"><p>the number of clusters.</p></dd><dt class="param">maxNeighbor</dt><dd class="cmt"><p>the maximum number of neighbors examined during a random search of local minima.</p></dd><dt class="param">numLocal</dt><dd class="cmt"><p>the number of local minima to search for.</p></dd></dl></div>
    </li><li name="smile.clustering#dac" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="dac(data:Array[Array[Double]],k:Int,alpha:Double,maxIter:Int,tol:Double,splitTol:Double):smile.clustering.DeterministicAnnealing"></a>
      <a id="dac(Array[Array[Double]],Int,Double,Int,Double,Double):DeterministicAnnealing"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">dac</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Double">Double</span>]]</span>, <span name="k">k: <span class="extype" name="scala.Int">Int</span></span>, <span name="alpha">alpha: <span class="extype" name="scala.Double">Double</span> = <span class="symbol">0.9</span></span>, <span name="maxIter">maxIter: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">100</span></span>, <span name="tol">tol: <span class="extype" name="scala.Double">Double</span> = <span class="symbol">1E-4</span></span>, <span name="splitTol">splitTol: <span class="extype" name="scala.Double">Double</span> = <span class="symbol">1E-2</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.DeterministicAnnealing">DeterministicAnnealing</span></span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@dac(data:Array[Array[Double]],k:Int,alpha:Double,maxIter:Int,tol:Double,splitTol:Double):smile.clustering.DeterministicAnnealing" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">Deterministic annealing clustering.</p><div class="fullcomment"><div class="comment cmt"><p>Deterministic annealing clustering. Deterministic annealing extends
soft-clustering to an annealing process.
For each temperature value, the algorithm iterates between the calculation
of all posteriori probabilities and the update of the centroids vectors,
until convergence is reached. The annealing starts with a high temperature.
Here, all centroids vectors converge to the center of the pattern
distribution (independent of their initial positions). Below a critical
temperature the vectors start to split. Further decreasing the temperature
leads to more splittings until all centroids vectors are separate. The
annealing can therefore avoid (if it is sufficiently slow) the convergence
to local minima.</p><h6>References:</h6><ul><li>Kenneth Rose. Deterministic Annealing for Clustering, Compression, Classification, Regression, and Speech Recognition.
</li></ul></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>the data set.</p></dd><dt class="param">k</dt><dd class="cmt"><p>the maximum number of clusters.</p></dd><dt class="param">alpha</dt><dd class="cmt"><p>the temperature T is decreasing as T = T * alpha. alpha has
                to be in (0, 1).</p></dd><dt class="param">tol</dt><dd class="cmt"><p>the tolerance of convergence test.</p></dd><dt class="param">splitTol</dt><dd class="cmt"><p>the tolerance to split a cluster.</p></dd></dl></div>
    </li><li name="smile.clustering#dbscan" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="dbscan(data:Array[Array[Double]],minPts:Int,radius:Double):smile.clustering.DBSCAN[Array[Double]]"></a>
      <a id="dbscan(Array[Array[Double]],Int,Double):DBSCAN[Array[Double]]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">dbscan</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Double">Double</span>]]</span>, <span name="minPts">minPts: <span class="extype" name="scala.Int">Int</span></span>, <span name="radius">radius: <span class="extype" name="scala.Double">Double</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.DBSCAN">DBSCAN</span>[<span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Double">Double</span>]]</span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@dbscan(data:Array[Array[Double]],minPts:Int,radius:Double):smile.clustering.DBSCAN[Array[Double]]" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">DBSCAN with Euclidean distance.</p><div class="fullcomment"><div class="comment cmt"><p>DBSCAN with Euclidean distance.
DBSCAN finds a number of clusters starting from the estimated density
distribution of corresponding nodes.
</p></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>the data set.</p></dd><dt class="param">minPts</dt><dd class="cmt"><p>the minimum number of neighbors for a core data point.</p></dd><dt class="param">radius</dt><dd class="cmt"><p>the neighborhood radius.</p></dd></dl></div>
    </li><li name="smile.clustering#dbscan" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="dbscan[T&lt;:AnyRef](data:Array[T],distance:smile.math.distance.Distance[T],minPts:Int,radius:Double):smile.clustering.DBSCAN[T]"></a>
      <a id="dbscan[T&lt;:AnyRef](Array[T],Distance[T],Int,Double):DBSCAN[T]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">dbscan</span><span class="tparams">[<span name="T">T &lt;: <span class="extype" name="scala.AnyRef">AnyRef</span></span>]</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="smile.clustering.dbscan.T">T</span>]</span>, <span name="distance">distance: <span class="extype" name="smile.math.distance.Distance">Distance</span>[<span class="extype" name="smile.clustering.dbscan.T">T</span>]</span>, <span name="minPts">minPts: <span class="extype" name="scala.Int">Int</span></span>, <span name="radius">radius: <span class="extype" name="scala.Double">Double</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.DBSCAN">DBSCAN</span>[<span class="extype" name="smile.clustering.dbscan.T">T</span>]</span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@dbscan[T&lt;:AnyRef](data:Array[T],distance:smile.math.distance.Distance[T],minPts:Int,radius:Double):smile.clustering.DBSCAN[T]" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">Density-Based Spatial Clustering of Applications with Noise.</p><div class="fullcomment"><div class="comment cmt"><p>Density-Based Spatial Clustering of Applications with Noise.
DBSCAN finds a number of clusters starting from the estimated density
distribution of corresponding nodes.
</p></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>the data set.</p></dd><dt class="param">distance</dt><dd class="cmt"><p>the distance metric.</p></dd><dt class="param">minPts</dt><dd class="cmt"><p>the minimum number of neighbors for a core data point.</p></dd><dt class="param">radius</dt><dd class="cmt"><p>the neighborhood radius.</p></dd></dl></div>
    </li><li name="smile.clustering#dbscan" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="dbscan[T&lt;:AnyRef](data:Array[T],nns:smile.neighbor.RNNSearch[T,T],minPts:Int,radius:Double):smile.clustering.DBSCAN[T]"></a>
      <a id="dbscan[T&lt;:AnyRef](Array[T],RNNSearch[T,T],Int,Double):DBSCAN[T]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">dbscan</span><span class="tparams">[<span name="T">T &lt;: <span class="extype" name="scala.AnyRef">AnyRef</span></span>]</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="smile.clustering.dbscan.T">T</span>]</span>, <span name="nns">nns: <span class="extype" name="smile.neighbor.RNNSearch">RNNSearch</span>[<span class="extype" name="smile.clustering.dbscan.T">T</span>, <span class="extype" name="smile.clustering.dbscan.T">T</span>]</span>, <span name="minPts">minPts: <span class="extype" name="scala.Int">Int</span></span>, <span name="radius">radius: <span class="extype" name="scala.Double">Double</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.DBSCAN">DBSCAN</span>[<span class="extype" name="smile.clustering.dbscan.T">T</span>]</span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@dbscan[T&lt;:AnyRef](data:Array[T],nns:smile.neighbor.RNNSearch[T,T],minPts:Int,radius:Double):smile.clustering.DBSCAN[T]" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">Density-Based Spatial Clustering of Applications with Noise.</p><div class="fullcomment"><div class="comment cmt"><p>Density-Based Spatial Clustering of Applications with Noise.
DBSCAN finds a number of clusters starting from the estimated density
distribution of corresponding nodes.</p><p>DBSCAN requires two parameters: radius (i.e. neighborhood radius) and the
number of minimum points required to form a cluster (minPts). It starts
with an arbitrary starting point that has not been visited. This point's
neighborhood is retrieved, and if it contains sufficient number of points,
a cluster is started. Otherwise, the point is labeled as noise. Note that
this point might later be found in a sufficiently sized radius-environment
of a different point and hence be made part of a cluster.</p><p>If a point is found to be part of a cluster, its neighborhood is also
part of that cluster. Hence, all points that are found within the
neighborhood are added, as is their own neighborhood. This process
continues until the cluster is completely found. Then, a new unvisited point
is retrieved and processed, leading to the discovery of a further cluster
of noise.</p><p>DBSCAN visits each point of the database, possibly multiple times (e.g.,
as candidates to different clusters). For practical considerations, however,
the time complexity is mostly governed by the number of nearest neighbor
queries. DBSCAN executes exactly one such query for each point, and if
an indexing structure is used that executes such a neighborhood query
in O(log n), an overall runtime complexity of O(n log n) is obtained.</p><p>DBSCAN has many advantages such as</p><ul><li>DBSCAN does not need to know the number of clusters in the data
a priori, as opposed to k-means.</li><li>DBSCAN can find arbitrarily shaped clusters. It can even find clusters
completely surrounded by (but not connected to) a different cluster.
Due to the MinPts parameter, the so-called single-link effect
(different clusters being connected by a thin line of points) is reduced.</li><li>DBSCAN has a notion of noise.</li><li>DBSCAN requires just two parameters and is mostly insensitive to the
ordering of the points in the database. (Only points sitting on the
edge of two different clusters might swap cluster membership if the
ordering of the points is changed, and the cluster assignment is unique
only up to isomorphism.)</li></ul><p>On the other hand, DBSCAN has the disadvantages of</p><ul><li>In high dimensional space, the data are sparse everywhere
because of the curse of dimensionality. Therefore, DBSCAN doesn't
work well on high-dimensional data in general.</li><li>DBSCAN does not respond well to data sets with varying densities.</li></ul><h6>References:</h6><ul><li>Martin Ester, Hans-Peter Kriegel, Jorg Sander, Xiaowei Xu (1996-). A density-based algorithm for discovering clusters in large spatial databases with noise&quot;. KDD, 1996.</li><li>Jorg Sander, Martin Ester, Hans-Peter  Kriegel, Xiaowei Xu. (1998). Density-Based Clustering in Spatial Databases: The Algorithm GDBSCAN and Its Applications. 1998.
</li></ul></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>the data set.</p></dd><dt class="param">nns</dt><dd class="cmt"><p>the data structure for neighborhood search.</p></dd><dt class="param">minPts</dt><dd class="cmt"><p>the minimum number of neighbors for a core data point.</p></dd><dt class="param">radius</dt><dd class="cmt"><p>the neighborhood radius.</p></dd></dl></div>
    </li><li name="smile.clustering#denclue" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="denclue(data:Array[Array[Double]],sigma:Double,m:Int):smile.clustering.DENCLUE"></a>
      <a id="denclue(Array[Array[Double]],Double,Int):DENCLUE"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">denclue</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Double">Double</span>]]</span>, <span name="sigma">sigma: <span class="extype" name="scala.Double">Double</span></span>, <span name="m">m: <span class="extype" name="scala.Int">Int</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.DENCLUE">DENCLUE</span></span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@denclue(data:Array[Array[Double]],sigma:Double,m:Int):smile.clustering.DENCLUE" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">DENsity CLUstering.</p><div class="fullcomment"><div class="comment cmt"><p>DENsity CLUstering. The DENCLUE algorithm employs a cluster model based on
kernel density estimation. A cluster is defined by a local maximum of the
estimated density function. Data points going to the same local maximum
are put into the same cluster.</p><p>Clearly, DENCLUE doesn't work on data with uniform distribution. In high
dimensional space, the data always look like uniformly distributed because
of the curse of dimensionality. Therefore, DENCLUDE doesn't work well on
high-dimensional data in general.</p><h6>References:</h6><ul><li>A. Hinneburg and D. A. Keim. A general approach to clustering in large databases with noise. Knowledge and Information Systems, 5(4):387-415, 2003.</li><li>Alexander Hinneburg and Hans-Henning Gabriel. DENCLUE 2.0: Fast Clustering based on Kernel Density Estimation. IDA, 2007.
</li></ul></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>the data set.</p></dd><dt class="param">sigma</dt><dd class="cmt"><p>the smooth parameter in the Gaussian kernel. The user can
             choose sigma such that number of density attractors is constant for a
             long interval of sigma.</p></dd><dt class="param">m</dt><dd class="cmt"><p>the number of selected samples used in the iteration.
             This number should be much smaller than the number of data points
             to speed up the algorithm. It should also be large enough to capture
             the sufficient information of underlying distribution.</p></dd></dl></div>
    </li><li name="smile.clustering#gmeans" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="gmeans(data:Array[Array[Double]],k:Int):smile.clustering.GMeans"></a>
      <a id="gmeans(Array[Array[Double]],Int):GMeans"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">gmeans</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Double">Double</span>]]</span>, <span name="k">k: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">100</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.GMeans">GMeans</span></span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@gmeans(data:Array[Array[Double]],k:Int):smile.clustering.GMeans" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">G-Means clustering algorithm, an extended K-Means which tries to
automatically determine the number of clusters by normality test.</p><div class="fullcomment"><div class="comment cmt"><p>G-Means clustering algorithm, an extended K-Means which tries to
automatically determine the number of clusters by normality test.
The G-means algorithm is based on a statistical test for the hypothesis
that a subset of data follows a Gaussian distribution. G-means runs
k-means with increasing k in a hierarchical fashion until the test accepts
the hypothesis that the data assigned to each k-means center are Gaussian.</p><h6>References:</h6><ul><li>G. Hamerly and C. Elkan. Learning the k in k-means. NIPS, 2003.
</li></ul></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>the data set.</p></dd><dt class="param">k</dt><dd class="cmt"><p>the maximum number of clusters.</p></dd></dl></div>
    </li><li name="smile.clustering#hclust" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="hclust[T&lt;:AnyRef](data:Array[T],distance:smile.math.distance.Distance[T],method:String):smile.clustering.HierarchicalClustering"></a>
      <a id="hclust[T&lt;:AnyRef](Array[T],Distance[T],String):HierarchicalClustering"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">hclust</span><span class="tparams">[<span name="T">T &lt;: <span class="extype" name="scala.AnyRef">AnyRef</span></span>]</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="smile.clustering.hclust.T">T</span>]</span>, <span name="distance">distance: <span class="extype" name="smile.math.distance.Distance">Distance</span>[<span class="extype" name="smile.clustering.hclust.T">T</span>]</span>, <span name="method">method: <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.HierarchicalClustering">HierarchicalClustering</span></span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@hclust[T&lt;:AnyRef](data:Array[T],distance:smile.math.distance.Distance[T],method:String):smile.clustering.HierarchicalClustering" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">Agglomerative Hierarchical Clustering.</p><div class="fullcomment"><div class="comment cmt"><p>Agglomerative Hierarchical Clustering. This method
seeks to build a hierarchy of clusters in a bottom up approach: each
observation starts in its own cluster, and pairs of clusters are merged as
one moves up the hierarchy. The results of hierarchical clustering are
usually presented in a dendrogram.</p><p>In general, the merges are determined in a greedy manner. In order to decide
which clusters should be combined, a measure of dissimilarity between sets
of observations is required. In most methods of hierarchical clustering,
this is achieved by use of an appropriate metric, and a linkage criteria
which specifies the dissimilarity of sets as a function of the pairwise
distances of observations in the sets.</p><p>Hierarchical clustering has the distinct advantage that any valid measure
of distance can be used. In fact, the observations themselves are not
required: all that is used is a matrix of distances.</p><h4> References </h4><ul><li>David Eppstein. Fast hierarchical clustering and other applications of dynamic closest pairs. SODA 1998.
</li></ul></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>The data set.</p></dd><dt class="param">distance</dt><dd class="cmt"><p>the distance/dissimilarity measure.</p></dd><dt class="param">method</dt><dd class="cmt"><p>the agglomeration method to merge clusters. This should be one of
              &quot;single&quot;, &quot;complete&quot;, &quot;upgma&quot;, &quot;upgmc&quot;, &quot;wpgma&quot;, &quot;wpgmc&quot;, and &quot;ward&quot;.</p></dd></dl></div>
    </li><li name="smile.clustering#hclust" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="hclust(data:Array[Array[Double]],method:String):smile.clustering.HierarchicalClustering"></a>
      <a id="hclust(Array[Array[Double]],String):HierarchicalClustering"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">hclust</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Double">Double</span>]]</span>, <span name="method">method: <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.HierarchicalClustering">HierarchicalClustering</span></span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@hclust(data:Array[Array[Double]],method:String):smile.clustering.HierarchicalClustering" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">Agglomerative Hierarchical Clustering.</p><div class="fullcomment"><div class="comment cmt"><p>Agglomerative Hierarchical Clustering. This method
seeks to build a hierarchy of clusters in a bottom up approach: each
observation starts in its own cluster, and pairs of clusters are merged as
one moves up the hierarchy. The results of hierarchical clustering are
usually presented in a dendrogram.</p><p>In general, the merges are determined in a greedy manner. In order to decide
which clusters should be combined, a measure of dissimilarity between sets
of observations is required. In most methods of hierarchical clustering,
this is achieved by use of an appropriate metric, and a linkage criteria
which specifies the dissimilarity of sets as a function of the pairwise
distances of observations in the sets.</p><p>Hierarchical clustering has the distinct advantage that any valid measure
of distance can be used. In fact, the observations themselves are not
required: all that is used is a matrix of distances.</p><h4> References </h4><ul><li>David Eppstein. Fast hierarchical clustering and other applications of dynamic closest pairs. SODA 1998.
</li></ul></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>The data set.</p></dd><dt class="param">method</dt><dd class="cmt"><p>the agglomeration method to merge clusters. This should be one of
              &quot;single&quot;, &quot;complete&quot;, &quot;upgma&quot;, &quot;upgmc&quot;, &quot;wpgma&quot;, &quot;wpgmc&quot;, and &quot;ward&quot;.</p></dd></dl></div>
    </li><li name="smile.clustering#kmeans" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="kmeans(data:Array[Array[Double]],k:Int,maxIter:Int,tol:Double,runs:Int):smile.clustering.KMeans"></a>
      <a id="kmeans(Array[Array[Double]],Int,Int,Double,Int):KMeans"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">kmeans</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Double">Double</span>]]</span>, <span name="k">k: <span class="extype" name="scala.Int">Int</span></span>, <span name="maxIter">maxIter: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">100</span></span>, <span name="tol">tol: <span class="extype" name="scala.Double">Double</span> = <span class="symbol">1E-4</span></span>, <span name="runs">runs: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">16</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.KMeans">KMeans</span></span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@kmeans(data:Array[Array[Double]],k:Int,maxIter:Int,tol:Double,runs:Int):smile.clustering.KMeans" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">K-Means clustering.</p><div class="fullcomment"><div class="comment cmt"><p>K-Means clustering. The algorithm partitions n observations into k clusters in which
each observation belongs to the cluster with the nearest mean.
Although finding an exact solution to the k-means problem for arbitrary
input is NP-hard, the standard approach to finding an approximate solution
(often called Lloyd's algorithm or the k-means algorithm) is used widely
and frequently finds reasonable solutions quickly.</p><p>However, the k-means algorithm has at least two major theoretic shortcomings:</p><ul><li>First, it has been shown that the worst case running time of the
algorithm is super-polynomial in the input size.</li><li>Second, the approximation found can be arbitrarily bad with respect
to the objective function compared to the optimal learn.</li></ul><p>In this implementation, we use k-means++ which addresses the second of these
obstacles by specifying a procedure to initialize the cluster centers before
proceeding with the standard k-means optimization iterations. With the
k-means++ initialization, the algorithm is guaranteed to find a solution
that is O(log k) competitive to the optimal k-means solution.</p><p>We also use k-d trees to speed up each k-means step as described in the filter
algorithm by Kanungo, et al.</p><p>K-means is a hard clustering method, i.e. each sample is assigned to
a specific cluster. In contrast, soft clustering, e.g. the
Expectation-Maximization algorithm for Gaussian mixtures, assign samples
to different clusters with different probabilities.</p><h6>References:</h6><ul><li>Tapas Kanungo, David M. Mount, Nathan S. Netanyahu, Christine D. Piatko, Ruth Silverman, and Angela Y. Wu. An Efficient k-Means Clustering Algorithm: Analysis and Implementation. IEEE TRANS. PAMI, 2002.</li><li>D. Arthur and S. Vassilvitskii. &quot;K-means++: the advantages of careful seeding&quot;. ACM-SIAM symposium on Discrete algorithms, 1027-1035, 2007.</li><li>Anna D. Peterson, Arka P. Ghosh and Ranjan Maitra. A systematic evaluation of different methods for initializing the K-means clustering algorithm. 2010.</li></ul><p>This method runs the algorithm for given times and return the best one with smallest distortion.
</p></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>the data set.</p></dd><dt class="param">k</dt><dd class="cmt"><p>the number of clusters.</p></dd><dt class="param">maxIter</dt><dd class="cmt"><p>the maximum number of iterations for each running.</p></dd><dt class="param">tol</dt><dd class="cmt"><p>the tolerance of convergence test.</p></dd><dt class="param">runs</dt><dd class="cmt"><p>the number of runs of K-Means algorithm.</p></dd></dl></div>
    </li><li name="smile.clustering#kmodes" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="kmodes(data:Array[Array[Int]],k:Int,maxIter:Int,runs:Int):smile.clustering.KModes"></a>
      <a id="kmodes(Array[Array[Int]],Int,Int,Int):KModes"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">kmodes</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Int">Int</span>]]</span>, <span name="k">k: <span class="extype" name="scala.Int">Int</span></span>, <span name="maxIter">maxIter: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">100</span></span>, <span name="runs">runs: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">10</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.KModes">KModes</span></span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@kmodes(data:Array[Array[Int]],k:Int,maxIter:Int,runs:Int):smile.clustering.KModes" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">K-Modes clustering.</p><div class="fullcomment"><div class="comment cmt"><p>K-Modes clustering. K-Modes is the binary equivalent for K-Means.
The mean update for centroids is replace by the mode one which is
a majority vote among element of each cluster.
</p></div></div>
    </li><li name="smile.clustering#mec" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="mec[T&lt;:AnyRef](data:Array[T],nns:smile.neighbor.RNNSearch[T,T],k:Int,radius:Double,y:Array[Int],tol:Double):smile.clustering.MEC[T]"></a>
      <a id="mec[T&lt;:AnyRef](Array[T],RNNSearch[T,T],Int,Double,Array[Int],Double):MEC[T]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">mec</span><span class="tparams">[<span name="T">T &lt;: <span class="extype" name="scala.AnyRef">AnyRef</span></span>]</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="smile.clustering.mec.T">T</span>]</span>, <span name="nns">nns: <span class="extype" name="smile.neighbor.RNNSearch">RNNSearch</span>[<span class="extype" name="smile.clustering.mec.T">T</span>, <span class="extype" name="smile.clustering.mec.T">T</span>]</span>, <span name="k">k: <span class="extype" name="scala.Int">Int</span></span>, <span name="radius">radius: <span class="extype" name="scala.Double">Double</span></span>, <span name="y">y: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Int">Int</span>]</span>, <span name="tol">tol: <span class="extype" name="scala.Double">Double</span> = <span class="symbol">1E-4</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.MEC">MEC</span>[<span class="extype" name="smile.clustering.mec.T">T</span>]</span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@mec[T&lt;:AnyRef](data:Array[T],nns:smile.neighbor.RNNSearch[T,T],k:Int,radius:Double,y:Array[Int],tol:Double):smile.clustering.MEC[T]" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">Nonparametric Minimum Conditional Entropy Clustering.</p><div class="fullcomment"><div class="comment cmt"><p>Nonparametric Minimum Conditional Entropy Clustering.
</p></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>the data set.</p></dd><dt class="param">nns</dt><dd class="cmt"><p>the data structure for neighborhood search.</p></dd><dt class="param">k</dt><dd class="cmt"><p>the number of clusters. Note that this is just a hint. The final
              number of clusters may be less.</p></dd><dt class="param">radius</dt><dd class="cmt"><p>the neighborhood radius.</p></dd><dt class="param">tol</dt><dd class="cmt"><p>the tolerance of convergence test.</p></dd></dl></div>
    </li><li name="smile.clustering#mec" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="mec(data:Array[Array[Double]],k:Int,radius:Double):smile.clustering.MEC[Array[Double]]"></a>
      <a id="mec(Array[Array[Double]],Int,Double):MEC[Array[Double]]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">mec</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Double">Double</span>]]</span>, <span name="k">k: <span class="extype" name="scala.Int">Int</span></span>, <span name="radius">radius: <span class="extype" name="scala.Double">Double</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.MEC">MEC</span>[<span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Double">Double</span>]]</span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@mec(data:Array[Array[Double]],k:Int,radius:Double):smile.clustering.MEC[Array[Double]]" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">Nonparametric Minimum Conditional Entropy Clustering.</p><div class="fullcomment"><div class="comment cmt"><p>Nonparametric Minimum Conditional Entropy Clustering. Assume Euclidean distance.
</p></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>the data set.</p></dd><dt class="param">k</dt><dd class="cmt"><p>the number of clusters. Note that this is just a hint. The final
              number of clusters may be less.</p></dd><dt class="param">radius</dt><dd class="cmt"><p>the neighborhood radius.</p></dd></dl></div>
    </li><li name="smile.clustering#mec" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="mec[T&lt;:AnyRef](data:Array[T],distance:smile.math.distance.Metric[T],k:Int,radius:Double):smile.clustering.MEC[T]"></a>
      <a id="mec[T&lt;:AnyRef](Array[T],Metric[T],Int,Double):MEC[T]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">mec</span><span class="tparams">[<span name="T">T &lt;: <span class="extype" name="scala.AnyRef">AnyRef</span></span>]</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="smile.clustering.mec.T">T</span>]</span>, <span name="distance">distance: <span class="extype" name="smile.math.distance.Metric">Metric</span>[<span class="extype" name="smile.clustering.mec.T">T</span>]</span>, <span name="k">k: <span class="extype" name="scala.Int">Int</span></span>, <span name="radius">radius: <span class="extype" name="scala.Double">Double</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.MEC">MEC</span>[<span class="extype" name="smile.clustering.mec.T">T</span>]</span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@mec[T&lt;:AnyRef](data:Array[T],distance:smile.math.distance.Metric[T],k:Int,radius:Double):smile.clustering.MEC[T]" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">Nonparametric Minimum Conditional Entropy Clustering.</p><div class="fullcomment"><div class="comment cmt"><p>Nonparametric Minimum Conditional Entropy Clustering.
</p></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>the data set.</p></dd><dt class="param">distance</dt><dd class="cmt"><p>the distance measure for neighborhood search.</p></dd><dt class="param">k</dt><dd class="cmt"><p>the number of clusters. Note that this is just a hint. The final
                number of clusters may be less.</p></dd><dt class="param">radius</dt><dd class="cmt"><p>the neighborhood radius.</p></dd></dl></div>
    </li><li name="smile.clustering#mec" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="mec[T&lt;:AnyRef](data:Array[T],distance:smile.math.distance.Distance[T],k:Int,radius:Double):smile.clustering.MEC[T]"></a>
      <a id="mec[T&lt;:AnyRef](Array[T],Distance[T],Int,Double):MEC[T]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">mec</span><span class="tparams">[<span name="T">T &lt;: <span class="extype" name="scala.AnyRef">AnyRef</span></span>]</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="smile.clustering.mec.T">T</span>]</span>, <span name="distance">distance: <span class="extype" name="smile.math.distance.Distance">Distance</span>[<span class="extype" name="smile.clustering.mec.T">T</span>]</span>, <span name="k">k: <span class="extype" name="scala.Int">Int</span></span>, <span name="radius">radius: <span class="extype" name="scala.Double">Double</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.MEC">MEC</span>[<span class="extype" name="smile.clustering.mec.T">T</span>]</span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@mec[T&lt;:AnyRef](data:Array[T],distance:smile.math.distance.Distance[T],k:Int,radius:Double):smile.clustering.MEC[T]" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">Nonparametric Minimum Conditional Entropy Clustering.</p><div class="fullcomment"><div class="comment cmt"><p>Nonparametric Minimum Conditional Entropy Clustering. This method performs
very well especially when the exact number of clusters is unknown.
The method can also correctly reveal the structure of data and effectively
identify outliers simultaneously.</p><p>The clustering criterion is based on the conditional entropy H(C | x), where
C is the cluster label and x is an observation. According to Fano's
inequality, we can estimate C with a low probability of error only if the
conditional entropy H(C | X) is small. MEC also generalizes the criterion
by replacing Shannon's entropy with Havrda-Charvat's structural
&alpha;-entropy. Interestingly, the minimum entropy criterion based
on structural &alpha;-entropy is equal to the probability error of the
nearest neighbor method when &alpha;= 2. To estimate p(C | x), MEC employs
Parzen density estimation, a nonparametric approach.</p><p>MEC is an iterative algorithm starting with an initial partition given by
any other clustering methods, e.g. k-means, CLARNAS, hierarchical clustering,
etc. Note that a random initialization is NOT appropriate.</p><h6>References:</h6><ul><li>Haifeng Li. All rights reserved., Keshu Zhang, and Tao Jiang. Minimum Entropy Clustering and Applications to Gene Expression Analysis. CSB, 2004.
</li></ul></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>the data set.</p></dd><dt class="param">distance</dt><dd class="cmt"><p>the distance measure for neighborhood search.</p></dd><dt class="param">k</dt><dd class="cmt"><p>the number of clusters. Note that this is just a hint. The final
                number of clusters may be less.</p></dd><dt class="param">radius</dt><dd class="cmt"><p>the neighborhood radius.</p></dd></dl></div>
    </li><li name="smile.clustering#sib" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="sib(data:Array[smile.util.SparseArray],k:Int,maxIter:Int,runs:Int):smile.clustering.SIB"></a>
      <a id="sib(Array[SparseArray],Int,Int,Int):SIB"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">sib</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="smile.util.SparseArray">SparseArray</span>]</span>, <span name="k">k: <span class="extype" name="scala.Int">Int</span></span>, <span name="maxIter">maxIter: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">100</span></span>, <span name="runs">runs: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">8</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.SIB">SIB</span></span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@sib(data:Array[smile.util.SparseArray],k:Int,maxIter:Int,runs:Int):smile.clustering.SIB" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">The Sequential Information Bottleneck algorithm.</p><div class="fullcomment"><div class="comment cmt"><p>The Sequential Information Bottleneck algorithm. SIB clusters co-occurrence
data such as text documents vs words. SIB is guaranteed to converge to a local
maximum of the information. Moreover, the time and space complexity are
significantly improved in contrast to the agglomerative IB algorithm.</p><p>In analogy to K-Means, SIB's update formulas are essentially same as the
EM algorithm for estimating finite Gaussian mixture model by replacing
regular Euclidean distance with Kullback-Leibler divergence, which is
clearly a better dissimilarity measure for co-occurrence data. However,
the common batch updating rule (assigning all instances to nearest centroids
and then updating centroids) of K-Means won't work in SIB, which has
to work in a sequential way (reassigning (if better) each instance then
immediately update related centroids). It might be because K-L divergence
is very sensitive and the centroids may be significantly changed in each
iteration in batch updating rule.</p><p>Note that this implementation has a little difference from the original
paper, in which a weighted Jensen-Shannon divergence is employed as a
criterion to assign a randomly-picked sample to a different cluster.
However, this doesn't work well in some cases as we experienced probably
because the weighted JS divergence gives too much weight to clusters which
is much larger than a single sample. In this implementation, we instead
use the regular/unweighted Jensen-Shannon divergence.</p><h6>References:</h6><ul><li>N. Tishby, F.C. Pereira, and W. Bialek. The information bottleneck method. 1999.</li><li>N. Slonim, N. Friedman, and N. Tishby. Unsupervised document classification using sequential information maximization. ACM SIGIR, 2002.</li><li>Jaakko Peltonen, Janne Sinkkonen, and Samuel Kaski. Sequential information bottleneck for finite data. ICML, 2004.
</li></ul></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>the data set.</p></dd><dt class="param">k</dt><dd class="cmt"><p>the number of clusters.</p></dd><dt class="param">maxIter</dt><dd class="cmt"><p>the maximum number of iterations.</p></dd><dt class="param">runs</dt><dd class="cmt"><p>the number of runs of SIB algorithm.</p></dd></dl></div>
    </li><li name="smile.clustering#specc" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="specc(data:Array[Array[Double]],k:Int,l:Int,sigma:Double):smile.clustering.SpectralClustering"></a>
      <a id="specc(Array[Array[Double]],Int,Int,Double):SpectralClustering"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">specc</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Double">Double</span>]]</span>, <span name="k">k: <span class="extype" name="scala.Int">Int</span></span>, <span name="l">l: <span class="extype" name="scala.Int">Int</span></span>, <span name="sigma">sigma: <span class="extype" name="scala.Double">Double</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.SpectralClustering">SpectralClustering</span></span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@specc(data:Array[Array[Double]],k:Int,l:Int,sigma:Double):smile.clustering.SpectralClustering" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">Spectral clustering with Nystrom approximation.</p><div class="fullcomment"><div class="comment cmt"><p>Spectral clustering with Nystrom approximation.
</p></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>the dataset for clustering.</p></dd><dt class="param">k</dt><dd class="cmt"><p>the number of clusters.</p></dd><dt class="param">l</dt><dd class="cmt"><p>the number of random samples for Nystrom approximation.</p></dd><dt class="param">sigma</dt><dd class="cmt"><p>the smooth/width parameter of Gaussian kernel, which
             is a somewhat sensitive parameter. To search for the best setting,
             one may pick the value that gives the tightest clusters (smallest
             distortion, see { @link #distortion()}) in feature space.</p></dd></dl></div>
    </li><li name="smile.clustering#specc" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="specc(data:Array[Array[Double]],k:Int,sigma:Double):smile.clustering.SpectralClustering"></a>
      <a id="specc(Array[Array[Double]],Int,Double):SpectralClustering"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">specc</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Double">Double</span>]]</span>, <span name="k">k: <span class="extype" name="scala.Int">Int</span></span>, <span name="sigma">sigma: <span class="extype" name="scala.Double">Double</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.SpectralClustering">SpectralClustering</span></span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@specc(data:Array[Array[Double]],k:Int,sigma:Double):smile.clustering.SpectralClustering" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">Spectral clustering.</p><div class="fullcomment"><div class="comment cmt"><p>Spectral clustering.
</p></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>the dataset for clustering.</p></dd><dt class="param">k</dt><dd class="cmt"><p>the number of clusters.</p></dd><dt class="param">sigma</dt><dd class="cmt"><p>the smooth/width parameter of Gaussian kernel, which
             is a somewhat sensitive parameter. To search for the best setting,
             one may pick the value that gives the tightest clusters (smallest
             distortion, see { @link #distortion()}) in feature space.</p></dd></dl></div>
    </li><li name="smile.clustering#specc" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="specc(W:smile.math.matrix.Matrix,k:Int):smile.clustering.SpectralClustering"></a>
      <a id="specc(Matrix,Int):SpectralClustering"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">specc</span><span class="params">(<span name="W">W: <span class="extype" name="smile.math.matrix.Matrix">Matrix</span></span>, <span name="k">k: <span class="extype" name="scala.Int">Int</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.SpectralClustering">SpectralClustering</span></span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@specc(W:smile.math.matrix.Matrix,k:Int):smile.clustering.SpectralClustering" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">Spectral Clustering.</p><div class="fullcomment"><div class="comment cmt"><p>Spectral Clustering. Given a set of data points, the similarity matrix may
be defined as a matrix S where S<sub>ij</sub> represents a measure of the
similarity between points. Spectral clustering techniques make use of the
spectrum of the similarity matrix of the data to perform dimensionality
reduction for clustering in fewer dimensions. Then the clustering will
be performed in the dimension-reduce space, in which clusters of non-convex
shape may become tight. There are some intriguing similarities between
spectral clustering methods and kernel PCA, which has been empirically
observed to perform clustering.</p><h6>References:</h6><ul><li>A.Y. Ng, M.I. Jordan, and Y. Weiss. On Spectral Clustering: Analysis and an algorithm. NIPS, 2001.</li><li>Marina Maila and Jianbo Shi. Learning segmentation by random walks. NIPS, 2000.</li><li>Deepak Verma and Marina Meila. A Comparison of Spectral Clustering Algorithms. 2003.
</li></ul></div><dl class="paramcmts block"><dt class="param">W</dt><dd class="cmt"><p>the adjacency matrix of graph.</p></dd><dt class="param">k</dt><dd class="cmt"><p>the number of clusters.</p></dd></dl></div>
    </li><li name="smile.clustering#xmeans" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="xmeans(data:Array[Array[Double]],k:Int):smile.clustering.XMeans"></a>
      <a id="xmeans(Array[Array[Double]],Int):XMeans"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">xmeans</span><span class="params">(<span name="data">data: <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Double">Double</span>]]</span>, <span name="k">k: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">100</span></span>)</span><span class="result">: <span class="extype" name="smile.clustering.XMeans">XMeans</span></span>
      </span>
      </h4><span class="permalink">
      <a href="../../index.html#smile.clustering.package@xmeans(data:Array[Array[Double]],k:Int):smile.clustering.XMeans" title="Permalink" target="_top">
        <img src="../../lib/permalink.png" alt="Permalink" />
      </a>
    </span>
      <p class="shortcomment cmt">X-Means clustering algorithm, an extended K-Means which tries to
automatically determine the number of clusters based on BIC scores.</p><div class="fullcomment"><div class="comment cmt"><p>X-Means clustering algorithm, an extended K-Means which tries to
automatically determine the number of clusters based on BIC scores.
Starting with only one cluster, the X-Means algorithm goes into action
after each run of K-Means, making local decisions about which subset of the
current centroids should split themselves in order to better fit the data.
The splitting decision is done by computing the Bayesian Information
Criterion (BIC).</p><h6>References:</h6><ul><li>Dan Pelleg and Andrew Moore. X-means: Extending K-means with Efficient Estimation of the Number of Clusters. ICML, 2000.
</li></ul></div><dl class="paramcmts block"><dt class="param">data</dt><dd class="cmt"><p>the data set.</p></dd><dt class="param">k</dt><dd class="cmt"><p>the maximum number of clusters.</p></dd></dl></div>
    </li></ol>
            </div>

        

        
        </div>

        <div id="inheritedMembers">
        <div class="parent" name="scala.AnyRef">
              <h3>Inherited from <span class="extype" name="scala.AnyRef">AnyRef</span></h3>
            </div><div class="parent" name="scala.Any">
              <h3>Inherited from <span class="extype" name="scala.Any">Any</span></h3>
            </div>
        
        </div>

        <div id="groupedMembers">
        <div class="group" name="Ungrouped">
              <h3>Ungrouped</h3>
              
            </div>
        </div>

      </div>

      <div id="tooltip"></div>

      <div id="footer">  </div>


    </body>
      </html>
